README.md

```diff
@@ -473,87 +388,104 @@ The `s` field MUST contain a [Varsig] of the [CBOR] encoded `scope` field.
 ```json
 {
   "scope": [
-    {
-      "/": "bafyreihtmwju3okftpeuqe3x3ux5e7c2jescakwnoiyv45vnicke4kdxy4"
-    },
-    {
-      "/": "bafyreieuo63r3y2nuycaq4b3q2xvco3nprlxiwzcfp4cuupgaywat3z6mq"
-    }
+    {"/": "bafyreihtmwju3okftpeuqe3x3ux5e7c2jescakwnoiyv45vnicke4kdxy4"},
+    {"/": "bafyreieuo63r3y2nuycaq4b3q2xvco3nprlxiwzcfp4cuupgaywat3z6mq"}
   ],
-  "s": {
-    "/": {
-      "bytes": "7aEDQIJB8XXJ6hWbwu40fN4bq8+Zq8BxyybSWXatMVU3VsL+yzVYpeJqsEBQE5rNtUJefR5rRCNimKNZMJjA9/udZQQ"
-    }
-  }
+  "s": {"/": {"bytes": "7aEDQIJB8XXJ6hWbwu40fN4bq8+Zq8BxyybSWXatMVU3VsL+yzVYpeJqsEBQE5rNtUJefR5rRCNimKNZMJjA9/udZQQ"}}
 }
 ```
 
-# 5 Invocation
+# 5 Task
 
 As [noted in the introduction][lazy-vs-eager], there is a difference between a reference to a function and calling that function. The [Invocation] is an instruction to the [Executor] to perform enclosed [Task]. [Invocation]s are not executable until they have been provided provable authority (in form of UCANs in the `prf` field) and an [Authorization] (in the `auth` field) from the [Invoker].
 
 The `auth` field MUST be contain an [Authorization] which signs over the `&Task` in `run`.
 
-Concretely, this means that the `&Task` MUST be present in the associated `auth`'s `scope` field. An `Invocation` where the associated [Authorization] does not include the [Task] in the `scope` MUST be considered invalid.
+Concretely, this means that the `&Task` MUST be present in the associated `auth`'s `scope` field. A `Receipt` where the associated [Authorization] does not include the [Task] in the `scope` MUST be considered invalid.
 
-## 5.1 Schema
+## 5.1 Task
 
-```ipldsch
-type Invocation struct {
-  v       SemVer
+```
+type Task struct {
+  run     &Instruction
+  meta    {String : Any}
+  prf     [&UCAN]
 
-  run     &Task
   # Receipt of the invocation that caused this invocation
-  cause   optional &Invocation
+  cause   optional &Receipt
```

Comment:
Will this create a recursive link chain between the issued in fx Task and the Reciept it's been issued from?
```
Reciept -> Result -> Task
      ^              |
      ^--------------
      ```
Worrying, because such a chain seems to be impossible to express in IPLD

Reply:
Perhaps one way to preserve causality yet have content-addresing is to introduce a separate Fx Reciept, from which joined Invocations can refer to Result Reciept
![image](https://github.com/ucan-wg/invocation/assets/13056889/8449cad4-8665-411f-ae54-8485d33a71c8)

Reply:
Perhaps yet another way is to refer from Fx to Instruction, and capture causality in Invocation of that Instruction

![image](https://github.com/ucan-wg/invocation/assets/13056889/4077bd19-76ea-47b1-ae84-73dc76c85f52)

Reply:
Another iteration of thoughts,
seems enough to:
1. have Instructions as Effects
2. capture causality in Tasks (as it is now)

![image](https://github.com/ucan-wg/invocation/assets/13056889/053849d5-6fd9-4642-bccc-5cf48a9e9bd7)

Let me know if it's bonkers :smile:

Reply:
To find a solution we could start from defining the problem.
From my understanding, we need a way to capture causality of subsequent invocations
From that standpoint, it seems causality could be kept at Invocation level, and Instruction that outputs Effects does not actually care about how it's configured (Task) and invoked, Invoker's in control of that.
So, perhaps we can have:
1. `Effect`s to be Instructions
2. How such Instructions is to be configured may be specified by the Invoker (in Authorization of the caused Invocation?), in the form of UCAN that authorizes Executor to issue further Tasks (with configuration, e.g., limiting gas) on behalf of Invoker.
3. How it's to be invoked may be specified in the same way
4. Executor issues Tasks and their Invocation on behalf of Invoker
5. Causality's captured at the level of Tasks, in the same way it is now

![image](https://github.com/ucan-wg/invocation/assets/13056889/ed7f66e1-2990-43ed-8038-0899057356cc)
`Effect`s depend on some run-time values from Instruction, but these values been baked-in into them at run-time, when they've been created by the Instruction.

---

There are other ways to provide `Effect`s with Result (note: Result is a subset of run-time values, which's meh already), but they'll only add complexity. Feel free to skip this section.
But if you're curious how we could complicate lives of Invoker and Executioner :smile:, we could either:
1) requiring Invoker to create effects as a function of Invocation, producing `Effect`s with await/ok on Instruction and require Executor to resolve it to the exact Receipt.
2) by requiring to create an Effect as a function of Result, baking it into `Effect`s.

These two seems to be redundant hussle.

They'd look like:
1)
![image](https://github.com/ucan-wg/invocation/assets/13056889/2bf2e78c-6a8b-48d2-9e2c-17e9ae1ec32f)

2) Here there's more work for the Executor, as it needs to specify the exact Receipt to be awaited somehow (and the spec needs to be extended to require that), and then, subsequently, he (or his delegatee) needs to resolve that value.
Another con is that, likely, it'll be of value for Effect-producing logic to have at least Result value (ideally run-time data, but we're short on that at this stage), 
![image](https://github.com/ucan-wg/invocation/assets/13056889/7e9c3dd9-879c-4359-a13a-1fe064f32f8e)
Here, the spec needs to be extended to require Executor to resolve await/ok to the `cause` Receipt.

Reply:
It's a good thing to raise, but it's not possible to create a cycle because of hashing. You'd have to find a hash cycle in SHA2-256, which is impractical.

You can emit a new copy of the invocation to emulate loops via recursion, but each will need a unique nonce and thus unique hash. This enqueuing behaviour lets a runtime get between each run and check if it wants to keep running, suspend, or move to a different machine, all while maintaining totality in each run.

Reply:
Correct that cycles are impractical to have, and are not intended.
My worry is that the spec atm implies a cycle.
```
Receipt -> Effect -> Invocation
  ^                   |
  ^---------cause-----
```
Since [Effects are Invocations](https://github.com/ucan-wg/invocation/pull/21/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5L780) that capture in `cause` the Receipt that caused them.

The suggested approach is to have Effects as Instructions. Then we won't have the cycle and it seems more semantically aligned with what Invoker means to express, I imagine he'd like to issue further Instructions and not care to create and invoke a task from within the caused Instruction (also that would require to have it's private key baked into the caused Instruction, as he would need to sign Task and Invocation), instead he issues Instruction and Executor creates Task and Invokation on Invoker's behalf.

Side-note, currently there is misalignment in spec about where `cause` resides, the commented line above has it under Task type, whereas in [here](https://github.com/ucan-wg/invocation/pull/21/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R422) it is described to reside under Invocation. Personally, `cause` at the Invocation level seems semantical.